File: app/src/main/java/wangdaye/com/geometricweather/weather/converters/CmaResultConverter.java
package wangdaye.com.geometricweather.weather.converters;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.TimeZone;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.basic.models.weather.Alert;
import wangdaye.com.geometricweather.common.basic.models.weather.Astro;
import wangdaye.com.geometricweather.common.basic.models.weather.Base;
import wangdaye.com.geometricweather.common.basic.models.weather.Current;
import wangdaye.com.geometricweather.common.basic.models.weather.Daily;
import wangdaye.com.geometricweather.common.basic.models.weather.HalfDay;
import wangdaye.com.geometricweather.common.basic.models.weather.Hourly;
import wangdaye.com.geometricweather.common.basic.models.weather.Precipitation;
import wangdaye.com.geometricweather.common.basic.models.weather.PrecipitationDuration;
import wangdaye.com.geometricweather.common.basic.models.weather.PrecipitationProbability;
import wangdaye.com.geometricweather.common.basic.models.weather.Temperature;
import wangdaye.com.geometricweather.common.basic.models.weather.UV;
import wangdaye.com.geometricweather.common.basic.models.weather.Weather;
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode;
import wangdaye.com.geometricweather.common.basic.models.weather.Wind;
import wangdaye.com.geometricweather.common.basic.models.weather.WindDegree;
import wangdaye.com.geometricweather.weather.json.cma.CmaWeatherResult;

/**
 * Converts the China Meteorological Administration (weather.cma.cn) responses into the
 * unified {@link Weather} model.
 *
 * Notes specific to CMA:
 * - The {@code now} block has no weather text/code, so the current condition is taken from
 *   today's daily day/night text.
 * - Daily wind has only a Chinese scale string ({@code 微风}/{@code 3级}), no numeric speed.
 * - Hourly forecast has no JSON endpoint; it is scraped from the {@code hour-table} HTML.
 *   The weather icon number in the HTML is mapped back to a {@link WeatherCode} via the
 *   {@code code -> text} pairs already present in the daily JSON (self-consistent).
 */
public class CmaResultConverter {

    private static final TimeZone CN_TZ = TimeZone.getTimeZone("Asia/Shanghai");

    @Nullable
    public static Weather convert(Context context, Location location,
                                  CmaWeatherResult result, @Nullable String hourlyHtml) {
        if (result == null || result.data == null) {
            return null;
        }
        try {
            CmaWeatherResult.Data data = result.data;

            String cityId = location.getCityId();
            if (TextUtils.isEmpty(cityId) && data.location != null) {
                cityId = data.location.id;
            }

            Map<Integer, WeatherCode> iconCodeMap = buildIconCodeMap(data.daily);

            return new Weather(
                    new Base(cityId, System.currentTimeMillis(), new Date(),
                            System.currentTimeMillis(), new Date(), System.currentTimeMillis()),
                    convertCurrent(context, data),
                    null,
                    convertDailyList(context, data),
                    convertHourlyList(context, hourlyHtml, iconCodeMap),
                    new ArrayList<>(),
                    convertAlertList(data)
            );
        } catch (Exception e) {
            return null;
        }
    }

    @NonNull
    private static Current convertCurrent(Context context, CmaWeatherResult.Data data) {
        CmaWeatherResult.Now now = data.now;

        // The now block carries no weather text; derive it from today's daily by day/night.
        boolean isDay = isDaytime();
        String weatherText = "";
        if (data.daily != null && !data.daily.isEmpty()) {
            CmaWeatherResult.Daily today = data.daily.get(0);
            weatherText = isDay ? today.dayText : today.nightText;
            if (TextUtils.isEmpty(weatherText)) {
                weatherText = isDay ? today.nightText : today.dayText;
            }
        }
        WeatherCode weatherCode = getWeatherCode(weatherText);

        int temp = now != null && now.temperature != null ? now.temperature.intValue() : 0;
        Integer feelsLike = now != null && now.feelst != null ? now.feelst.intValue() : null;
        Float precip = now != null ? toFloat(now.precipitation) : null;
        Float humidity = now != null ? toFloat(now.humidity) : null;
        Float pressure = now != null ? toFloat(now.pressure) : null;
        float windKph = now != null && now.windSpeed != null ? msToKph(now.windSpeed) : 0f;
        String windDir = now != null && now.windDirection != null ? now.windDirection : "";
        float windDegree = now != null && now.windDirectionDegree != null
                ? now.windDirectionDegree.floatValue() : 0f;
        boolean noWindDir = TextUtils.isEmpty(windDir) || windDir.contains("无");

        return new Current(
                weatherText != null ? weatherText : "",
                weatherCode,
                new Temperature(temp, feelsLike, null, null, null, null, null),
                new Precipitation(precip, null, null, null, null),
                new PrecipitationProbability(null, null, null, null, null),
                new Wind(windDir, new WindDegree(windDegree, noWindDir), windKph,
                        CommonConverter.getWindLevel(context, windKph)),
                new UV(null, null, null),
                null,
                humidity, pressure, null, null, null, null, null, null
        );
    }

    @NonNull
    private static List<Daily> convertDailyList(Context context, CmaWeatherResult.Data data) {
        List<Daily> list = new ArrayList<>();
        if (data.daily == null) {
            return list;
        }
        for (CmaWeatherResult.Daily d : data.daily) {
            Date date = parseDate(d.date);
            if (date == null) {
                continue;
            }
            int high = d.high != null ? d.high.intValue() : 0;
            int low = d.low != null ? d.low.intValue() : 0;

            HalfDay day = buildHalfDay(context, "Day", d.dayText, d.dayWindDirection,
                    d.dayWindScale, high);
            HalfDay night = buildHalfDay(context, "Night", d.nightText, d.nightWindDirection,
                    d.nightWindScale, low);

            list.add(new Daily(date, date.getTime(), day, night, null, null, null,
                    null, null, new UV(null, null, null), 0f));
        }
        return list;
    }

    @NonNull
    private static HalfDay buildHalfDay(Context context, String phase, String text,
                                        String windDirection, String windScale, int temp) {
        String windDir = windDirection != null ? windDirection : "";
        boolean noWindDir = TextUtils.isEmpty(windDir) || windDir.contains("无");
        // Daily wind has no numeric speed, only a Chinese scale string used as the display level.
        String level = !TextUtils.isEmpty(windScale) ? windScale
                : CommonConverter.getWindLevel(context, 0);
        return new HalfDay(
                text != null ? text : "",
                phase,
                getWeatherCode(text),
                new Temperature(temp, null, null, null, null, null, null),
                new Precipitation(null, null, null, null, null),
                new PrecipitationProbability(null, null, null, null, null),
                new PrecipitationDuration(null, null, null, null, null),
                new Wind(windDir, new WindDegree(0, noWindDir), 0f, level),
                null
        );
    }

    @NonNull
    private static List<Hourly> convertHourlyList(Context context, @Nullable String html,
                                                  Map<Integer, WeatherCode> iconCodeMap) {
        List<Hourly> list = new ArrayList<>();
        if (TextUtils.isEmpty(html)) {
            return list;
        }
        try {
            Pattern tablePattern = Pattern.compile(
                    "(?s)<table[^>]*class=\"hour-table\"[^>]*>(.*?)</table>");
            Matcher tableMatcher = tablePattern.matcher(html);

            Calendar cal = Calendar.getInstance(CN_TZ);
            int prevHour = -1;
            boolean started = false;

            while (tableMatcher.find()) {
                String table = tableMatcher.group(1);
                Map<String, List<String>> rows = parseRows(table);

                List<String> times = rows.get("时间");
                if (times == null || times.isEmpty()) {
                    continue;
                }
                List<String> weathers = rows.get("天气");
                List<String> temps = rows.get("气温");
                List<String> winds = rows.get("风速");
                List<String> windDirs = rows.get("风向");

                for (int i = 0; i < times.size(); i++) {
                    Integer hour = parseHour(times.get(i));
                    if (hour == null) {
                        continue;
                    }
                    if (!started) {
                        started = true;
                    } else if (hour <= prevHour) {
                        cal.add(Calendar.DAY_OF_MONTH, 1);
                    }
                    prevHour = hour;
                    cal.set(Calendar.HOUR_OF_DAY, hour);
                    cal.set(Calendar.MINUTE, 0);
                    cal.set(Calendar.SECOND, 0);
                    cal.set(Calendar.MILLISECOND, 0);
                    Date date = cal.getTime();

                    WeatherCode code = WeatherCode.CLEAR;
                    Integer icon = at(weathers, i) != null ? extractIcon(weathers.get(i)) : null;
                    if (icon != null && iconCodeMap.containsKey(icon)) {
                        code = iconCodeMap.get(icon);
                    }

                    Integer temp = parseNumber(at(temps, i));
                    float windKph = 0f;
                    Float ws = parseDecimal(at(winds, i));
                    if (ws != null) {
                        windKph = ws * 3.6f;
                    }
                    String wdir = at(windDirs, i) != null ? stripTags(windDirs.get(i)) : "";
                    boolean noWindDir = TextUtils.isEmpty(wdir) || wdir.contains("无");

                    list.add(new Hourly(
                            date, date.getTime(), hour >= 6 && hour < 18,
                            "", code,
                            new Temperature(temp != null ? temp : 0, null, null, null, null, null, null),
                            new Precipitation(null, null, null, null, null),
                            new PrecipitationProbability(null, null, null, null, null),
                            new Wind(wdir, new WindDegree(0, noWindDir), windKph,
                                    CommonConverter.getWindLevel(context, windKph)),
                            new UV(null, null, null)
                    ));
                }
            }
        } catch (Exception e) {
            return new ArrayList<>();
        }
        return list;
    }

    @NonNull
    private static List<Alert> convertAlertList(CmaWeatherResult.Data data) {
        List<Alert> list = new ArrayList<>();
        if (data.alarm == null) {
            return list;
        }
        long id = 0;
        for (CmaWeatherResult.Alarm a : data.alarm) {
            if (a == null) {
                continue;
            }
            String title = !TextUtils.isEmpty(a.headline) ? a.headline : a.title;
            if (TextUtils.isEmpty(title) && TextUtils.isEmpty(a.description)) {
                continue;
            }
            Date date = parseDateTime(a.effective);
            long time = date != null ? date.getTime() : System.currentTimeMillis();
            list.add(new Alert(
                    id++,
                    date != null ? date : new Date(),
                    time,
                    title != null ? title : "",
                    a.description != null ? a.description : "",
                    a.type != null ? a.type : "",
                    1,
                    0xFFFFB82B
            ));
        }
        return list;
    }

    // builds icon-number -> WeatherCode from the daily code/text pairs (self-consistent).
    @NonNull
    private static Map<Integer, WeatherCode> buildIconCodeMap(@Nullable List<CmaWeatherResult.Daily> daily) {
        Map<Integer, WeatherCode> map = new HashMap<>();
        if (daily == null) {
            return map;
        }
        for (CmaWeatherResult.Daily d : daily) {
            if (d.dayCode != null && !TextUtils.isEmpty(d.dayText)) {
                map.put(d.dayCode, getWeatherCode(d.dayText));
            }
            if (d.nightCode != null && !TextUtils.isEmpty(d.nightText)) {
                map.put(d.nightCode, getWeatherCode(d.nightText));
            }
        }
        return map;
    }

    @NonNull
    static WeatherCode getWeatherCode(@Nullable String text) {
        if (TextUtils.isEmpty(text)) {
            return WeatherCode.CLEAR;
        }
        if (text.contains("雷") && text.contains("雨")) {
            return WeatherCode.THUNDERSTORM;
        }
        if (text.contains("雷")) {
            return WeatherCode.THUNDER;
        }
        if (text.contains("雹")) {
            return WeatherCode.HAIL;
        }
        if (text.contains("雨夹雪") || text.contains("雨雪") || (text.contains("雨") && text.contains("雪"))) {
            return WeatherCode.SLEET;
        }
        if (text.contains("雪")) {
            return WeatherCode.SNOW;
        }
        if (text.contains("雨")) {
            return WeatherCode.RAIN;
        }
        if (text.contains("雾")) {
            return WeatherCode.FOG;
        }
        if (text.contains("霾") || text.contains("沙") || text.contains("尘")) {
            return WeatherCode.HAZE;
        }
        if (text.contains("风") || text.contains("飑") || text.contains("龙卷")) {
            return WeatherCode.WIND;
        }
        if (text.contains("阴")) {
            return WeatherCode.CLOUDY;
        }
        if (text.contains("多云")) {
            return WeatherCode.PARTLY_CLOUDY;
        }
        if (text.contains("晴")) {
            return WeatherCode.CLEAR;
        }
        return WeatherCode.CLEAR;
    }

    // ---- HTML helpers ----

    private static Map<String, List<String>> parseRows(String tableHtml) {
        Map<String, List<String>> rows = new HashMap<>();
        Matcher rowMatcher = Pattern.compile("(?s)<tr[^>]*>(.*?)</tr>").matcher(tableHtml);
        while (rowMatcher.find()) {
            List<String> cells = new ArrayList<>();
            Matcher cellMatcher = Pattern.compile("(?s)<td[^>]*>(.*?)</td>").matcher(rowMatcher.group(1));
            while (cellMatcher.find()) {
                cells.add(cellMatcher.group(1));
            }
            if (cells.isEmpty()) {
                continue;
            }
            String label = stripTags(cells.get(0));
            if (TextUtils.isEmpty(label)) {
                continue;
            }
            rows.put(label, cells.subList(1, cells.size()));
        }
        return rows;
    }

    @Nullable
    private static String at(@Nullable List<String> list, int i) {
        return list != null && i < list.size() ? list.get(i) : null;
    }

    @Nullable
    private static Integer extractIcon(String cellHtml) {
        Matcher m = Pattern.compile("w(\\d+)\\.png").matcher(cellHtml);
        if (m.find()) {
            try {
                return Integer.parseInt(m.group(1));
            } catch (NumberFormatException ignored) {
            }
        }
        return null;
    }

    private static String stripTags(String s) {
        if (s == null) {
            return "";
        }
        return s.replaceAll("(?s)<[^>]*>", "")
                .replace("&nbsp;", " ")
                .trim();
    }

    @Nullable
    private static Integer parseHour(String cell) {
        String t = stripTags(cell);
        Matcher m = Pattern.compile("(\\d{1,2}):\\d{2}").matcher(t);
        if (m.find()) {
            try {
                return Integer.parseInt(m.group(1));
            } catch (NumberFormatException ignored) {
            }
        }
        return null;
    }

    @Nullable
    private static Integer parseNumber(@Nullable String cell) {
        Float f = parseDecimal(cell);
        return f != null ? Math.round(f) : null;
    }

    @Nullable
    private static Float parseDecimal(@Nullable String cell) {
        if (cell == null) {
            return null;
        }
        Matcher m = Pattern.compile("-?\\d+(\\.\\d+)?").matcher(stripTags(cell));
        if (m.find()) {
            try {
                return Float.parseFloat(m.group());
            } catch (NumberFormatException ignored) {
            }
        }
        return null;
    }

    // ---- misc helpers ----

    private static boolean isDaytime() {
        int hour = Calendar.getInstance(CN_TZ).get(Calendar.HOUR_OF_DAY);
        return hour >= 6 && hour < 18;
    }

    @Nullable
    private static Float toFloat(@Nullable Double d) {
        return d != null ? d.floatValue() : null;
    }

    private static float msToKph(double ms) {
        return (float) (ms * 3.6);
    }

    @Nullable
    private static Date parseDate(@Nullable String s) {
        if (s == null) {
            return null;
        }
        try {
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy/MM/dd", Locale.US);
            sdf.setTimeZone(CN_TZ);
            return sdf.parse(s);
        } catch (ParseException e) {
            return null;
        }
    }

    @Nullable
    private static Date parseDateTime(@Nullable String s) {
        if (s == null) {
            return null;
        }
        try {
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy/MM/dd HH:mm", Locale.US);
            sdf.setTimeZone(CN_TZ);
            return sdf.parse(s);
        } catch (ParseException e) {
            return null;
        }
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/weather/converters/CmaResultConverter.kt
package wangdaye.com.geometricweather.weather.converters

import android.content.Context
import android.text.TextUtils
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.regex.Pattern
import wangdaye.com.geometricweather.common.basic.models.Location
import wangdaye.com.geometricweather.common.basic.models.weather.AirQuality
import wangdaye.com.geometricweather.common.basic.models.weather.Alert
import wangdaye.com.geometricweather.common.basic.models.weather.Base
import wangdaye.com.geometricweather.common.basic.models.weather.Current
import wangdaye.com.geometricweather.common.basic.models.weather.Daily
import wangdaye.com.geometricweather.common.basic.models.weather.HalfDay
import wangdaye.com.geometricweather.common.basic.models.weather.Hourly
import wangdaye.com.geometricweather.common.basic.models.weather.Minutely
import wangdaye.com.geometricweather.common.basic.models.weather.Precipitation
import wangdaye.com.geometricweather.common.basic.models.weather.PrecipitationDuration
import wangdaye.com.geometricweather.common.basic.models.weather.PrecipitationProbability
import wangdaye.com.geometricweather.common.basic.models.weather.Temperature
import wangdaye.com.geometricweather.common.basic.models.weather.UV
import wangdaye.com.geometricweather.common.basic.models.weather.Weather
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode
import wangdaye.com.geometricweather.common.basic.models.weather.Wind
import wangdaye.com.geometricweather.common.basic.models.weather.WindDegree
import wangdaye.com.geometricweather.weather.json.cma.CmaWeatherResult

/**
 * Converts the China Meteorological Administration (weather.cma.cn) responses into the unified
 * [Weather] model.
 *
 * Notes specific to CMA:
 * - The `now` block has no weather text/code, so the current condition is taken from today's daily
 *   day/night text.
 * - Daily wind has only a Chinese scale string (`微风`/`3级`), no numeric speed.
 * - Hourly forecast has no JSON endpoint; it is scraped from the `hour-table` HTML. The weather
 *   icon number in the HTML is mapped back to a [WeatherCode] via the `code -> text` pairs already
 *   present in the daily JSON (self-consistent).
 * - Fields off [CmaWeatherResult] are Java platform types: Kotlin will not force a null check on
 *   them, so the explicit guards below carry their weight and must stay.
 */
object CmaResultConverter {

    private val CN_TZ: TimeZone = TimeZone.getTimeZone("Asia/Shanghai")

    private val HOUR_TABLE = Pattern.compile("(?s)<table[^>]*class=\"hour-table\"[^>]*>(.*?)</table>")
    private val TABLE_ROW = Pattern.compile("(?s)<tr[^>]*>(.*?)</tr>")
    private val TABLE_CELL = Pattern.compile("(?s)<td[^>]*>(.*?)</td>")
    private val HTML_TAG = Regex("(?s)<[^>]*>")
    private val ICON_NUMBER = Pattern.compile("w(\\d+)\\.png")
    private val HOUR_MINUTE = Pattern.compile("(\\d{1,2}):\\d{2}")
    private val DECIMAL = Pattern.compile("-?\\d+(\\.\\d+)?")

    @JvmStatic
    fun convert(
        context: Context,
        location: Location,
        result: CmaWeatherResult?,
        hourlyHtml: String?
    ): Weather? {
        val data = result?.data ?: return null
        return try {
            var cityId: String? = location.cityId
            if (TextUtils.isEmpty(cityId) && data.location != null) {
                cityId = data.location.id
            }

            val iconCodeMap = buildIconCodeMap(data.daily)
            val now = System.currentTimeMillis()

            Weather(
                Base(cityId, now, Date(), now, Date(), now),
                convertCurrent(context, data),
                null,
                convertDailyList(context, data),
                convertHourlyList(context, hourlyHtml, iconCodeMap),
                ArrayList<Minutely>(),
                convertAlertList(data)
            )
        } catch (e: Exception) {
            null
        }
    }

    private fun convertCurrent(context: Context, data: CmaWeatherResult.Data): Current {
        val now = data.now

        // The now block carries no weather text; derive it from today's daily by day/night.
        val isDay = isDaytime()
        var weatherText: String? = ""
        if (!data.daily.isNullOrEmpty()) {
            val today = data.daily[0]
            weatherText = if (isDay) today.dayText else today.nightText
            if (TextUtils.isEmpty(weatherText)) {
                weatherText = if (isDay) today.nightText else today.dayText
            }
        }

        val windKph = now?.windSpeed?.let { msToKph(it) } ?: 0f
        val windDir = now?.windDirection ?: ""
        val noWindDir = TextUtils.isEmpty(windDir) || windDir.contains("无")

        return Current(
            weatherText ?: "",
            getWeatherCode(weatherText),
            Temperature(
                now?.temperature?.toInt() ?: 0,
                now?.feelst?.toInt(),
                null, null, null, null, null
            ),
            Precipitation(now?.precipitation?.toFloat(), null, null, null, null),
            PrecipitationProbability(null, null, null, null, null),
            Wind(
                windDir,
                WindDegree(now?.windDirectionDegree?.toFloat() ?: 0f, noWindDir),
                windKph,
                CommonConverter.getWindLevel(context, windKph.toDouble())
            ),
            UV(null, null, null),
            AirQuality(null, null, null, null, null, null, null, null),
            now?.humidity?.toFloat(),
            now?.pressure?.toFloat(),
            null, null, null, null, null, null
        )
    }

    private fun convertDailyList(context: Context, data: CmaWeatherResult.Data): List<Daily> {
        val list = ArrayList<Daily>()
        val dailyList = data.daily ?: return list

        for (d in dailyList) {
            val date = parseDate(d.date) ?: continue

            list.add(
                Daily(
                    date,
                    date.time,
                    buildHalfDay(
                        context, "Day", d.dayText, d.dayWindDirection, d.dayWindScale,
                        d.high?.toInt() ?: 0
                    ),
                    buildHalfDay(
                        context, "Night", d.nightText, d.nightWindDirection, d.nightWindScale,
                        d.low?.toInt() ?: 0
                    ),
                    null, null, null, null, null,
                    UV(null, null, null),
                    0f
                )
            )
        }
        return list
    }

    private fun buildHalfDay(
        context: Context,
        phase: String,
        text: String?,
        windDirection: String?,
        windScale: String?,
        temperature: Int
    ): HalfDay {
        val windDir = windDirection ?: ""
        val noWindDir = TextUtils.isEmpty(windDir) || windDir.contains("无")
        // Daily wind has no numeric speed, only a Chinese scale string used as the display level.
        val level = if (!TextUtils.isEmpty(windScale)) {
            windScale!!
        } else {
            CommonConverter.getWindLevel(context, 0.0)
        }
        return HalfDay(
            text ?: "",
            phase,
            getWeatherCode(text),
            Temperature(temperature, null, null, null, null, null, null),
            Precipitation(null, null, null, null, null),
            PrecipitationProbability(null, null, null, null, null),
            PrecipitationDuration(null, null, null, null, null),
            Wind(windDir, WindDegree(0f, noWindDir), 0f, level),
            null
        )
    }

    private fun convertHourlyList(
        context: Context,
        html: String?,
        iconCodeMap: Map<Int, WeatherCode>
    ): List<Hourly> {
        val list = ArrayList<Hourly>()
        if (html.isNullOrEmpty()) {
            return list
        }
        try {
            val tableMatcher = HOUR_TABLE.matcher(html)

            val cal = Calendar.getInstance(CN_TZ)
            var prevHour = -1
            var started = false

            while (tableMatcher.find()) {
                val rows = parseRows(tableMatcher.group(1) ?: continue)

                val times = rows["时间"]
                if (times.isNullOrEmpty()) {
                    continue
                }
                val weathers = rows["天气"]
                val temps = rows["气温"]
                val winds = rows["风速"]
                val windDirs = rows["风向"]

                for (i in times.indices) {
                    val hour = parseHour(times[i]) ?: continue

                    if (!started) {
                        started = true
                    } else if (hour <= prevHour) {
                        cal.add(Calendar.DAY_OF_MONTH, 1)
                    }
                    prevHour = hour
                    cal.set(Calendar.HOUR_OF_DAY, hour)
                    cal.set(Calendar.MINUTE, 0)
                    cal.set(Calendar.SECOND, 0)
                    cal.set(Calendar.MILLISECOND, 0)
                    val date = cal.time

                    var code = WeatherCode.CLEAR
                    val icon = at(weathers, i)?.let { extractIcon(it) }
                    if (icon != null) {
                        iconCodeMap[icon]?.let { code = it }
                    }

                    val windKph = (parseDecimal(at(winds, i)) ?: 0f) * 3.6f
                    val windDir = at(windDirs, i)?.let { stripTags(it) } ?: ""
                    val noWindDir = TextUtils.isEmpty(windDir) || windDir.contains("无")

                    list.add(
                        Hourly(
                            date,
                            date.time,
                            hour in 6..17,
                            "",
                            code,
                            Temperature(
                                parseNumber(at(temps, i)) ?: 0,
                                null, null, null, null, null, null
                            ),
                            Precipitation(null, null, null, null, null),
                            PrecipitationProbability(null, null, null, null, null),
                            Wind(
                                windDir,
                                WindDegree(0f, noWindDir),
                                windKph,
                                CommonConverter.getWindLevel(context, windKph.toDouble())
                            ),
                            UV(null, null, null)
                        )
                    )
                }
            }
        } catch (e: Exception) {
            return ArrayList()
        }
        return list
    }

    private fun convertAlertList(data: CmaWeatherResult.Data): List<Alert> {
        val list = ArrayList<Alert>()
        val alarms = data.alarm ?: return list

        var id = 0L
        for (a in alarms) {
            if (a == null) {
                continue
            }
            val title = if (!TextUtils.isEmpty(a.headline)) a.headline else a.title
            if (TextUtils.isEmpty(title) && TextUtils.isEmpty(a.description)) {
                continue
            }
            val date = parseDateTime(a.effective)
            list.add(
                Alert(
                    id++,
                    date ?: Date(),
                    date?.time ?: System.currentTimeMillis(),
                    title ?: "",
                    a.description ?: "",
                    a.type ?: "",
                    1,
                    0xFFFFB82B.toInt()
                )
            )
        }
        return list
    }

    /** Builds icon-number -> WeatherCode from the daily code/text pairs (self-consistent). */
    private fun buildIconCodeMap(daily: List<CmaWeatherResult.Daily>?): Map<Int, WeatherCode> {
        val map = HashMap<Int, WeatherCode>()
        if (daily == null) {
            return map
        }
        for (d in daily) {
            if (d.dayCode != null && !TextUtils.isEmpty(d.dayText)) {
                map[d.dayCode] = getWeatherCode(d.dayText)
            }
            if (d.nightCode != null && !TextUtils.isEmpty(d.nightText)) {
                map[d.nightCode] = getWeatherCode(d.nightText)
            }
        }
        return map
    }

    /** Also used by [ApihzResultConverter], which shares CMA's Chinese weather-text vocabulary. */
    internal fun getWeatherCode(text: String?): WeatherCode {
        if (text.isNullOrEmpty()) {
            return WeatherCode.CLEAR
        }
        return when {
            text.contains("雷") && text.contains("雨") -> WeatherCode.THUNDERSTORM
            text.contains("雷") -> WeatherCode.THUNDER
            text.contains("雹") -> WeatherCode.HAIL
            text.contains("雨夹雪") || text.contains("雨雪")
                    || (text.contains("雨") && text.contains("雪")) -> WeatherCode.SLEET
            text.contains("雪") -> WeatherCode.SNOW
            text.contains("雨") -> WeatherCode.RAIN
            text.contains("雾") -> WeatherCode.FOG
            text.contains("霾") || text.contains("沙")
                    || text.contains("尘") -> WeatherCode.HAZE
            text.contains("风") || text.contains("飑")
                    || text.contains("龙卷") -> WeatherCode.WIND
            text.contains("阴") -> WeatherCode.CLOUDY
            text.contains("多云") -> WeatherCode.PARTLY_CLOUDY
            text.contains("晴") -> WeatherCode.CLEAR
            else -> WeatherCode.CLEAR
        }
    }

    // ---- HTML helpers ----

    private fun parseRows(tableHtml: String): Map<String, List<String>> {
        val rows = HashMap<String, List<String>>()
        val rowMatcher = TABLE_ROW.matcher(tableHtml)
        while (rowMatcher.find()) {
            val cells = ArrayList<String>()
            val cellMatcher = TABLE_CELL.matcher(rowMatcher.group(1) ?: continue)
            while (cellMatcher.find()) {
                cellMatcher.group(1)?.let { cells.add(it) }
            }
            if (cells.isEmpty()) {
                continue
            }
            val label = stripTags(cells[0])
            if (TextUtils.isEmpty(label)) {
                continue
            }
            rows[label] = cells.subList(1, cells.size)
        }
        return rows
    }

    private fun at(list: List<String>?, i: Int): String? =
        if (list != null && i < list.size) list[i] else null

    private fun extractIcon(cellHtml: String): Int? {
        val m = ICON_NUMBER.matcher(cellHtml)
        if (m.find()) {
            return m.group(1)?.toIntOrNull()
        }
        return null
    }

    private fun stripTags(s: String?): String {
        if (s == null) {
            return ""
        }
        return s.replace(HTML_TAG, "")
            .replace("&nbsp;", " ")
            .trim()
    }

    private fun parseHour(cell: String): Int? {
        val m = HOUR_MINUTE.matcher(stripTags(cell))
        if (m.find()) {
            return m.group(1)?.toIntOrNull()
        }
        return null
    }

    private fun parseNumber(cell: String?): Int? = parseDecimal(cell)?.let { Math.round(it) }

    private fun parseDecimal(cell: String?): Float? {
        if (cell == null) {
            return null
        }
        val m = DECIMAL.matcher(stripTags(cell))
        if (m.find()) {
            return m.group().toFloatOrNull()
        }
        return null
    }

    // ---- misc helpers ----

    private fun isDaytime(): Boolean =
        Calendar.getInstance(CN_TZ).get(Calendar.HOUR_OF_DAY) in 6..17

    private fun msToKph(ms: Double): Float = (ms * 3.6).toFloat()

    private fun parseDate(s: String?): Date? = parse(s, "yyyy/MM/dd")

    private fun parseDateTime(s: String?): Date? = parse(s, "yyyy/MM/dd HH:mm")

    private fun parse(s: String?, pattern: String): Date? {
        if (s == null) {
            return null
        }
        return try {
            val sdf = SimpleDateFormat(pattern, Locale.US)
            sdf.timeZone = CN_TZ
            sdf.parse(s)
        } catch (e: ParseException) {
            null
        }
    }
}
